rusqlite_migration/src/loader.rs

```diff
@@ -95,50 +91,47 @@ impl<'u> From<&MigrationFile> for M<'u> {
 }
 
 pub(crate) fn from_directory(dir: &'static Dir<'static>) -> Result<Vec<M<'static>>> {
-    // We want to limit the number of allocations here
-    let mut btreemap = BTreeMap::new();
+    let mut migrations: Vec<Option<M>> = vec![None; dir.dirs().count() + 1];
 
     // We cannot use FromIterator<(K, V)> for BTreeMap<K, V, Global> because that currently
     // allocates a Vec and sorts it in the background, which we want to avoid
     for dir in dir.dirs() {
         let migration_file = MigrationFile::try_from(dir)?;
-        let entry = btreemap.entry(migration_file.id);
 
-        if let Entry::Occupied(_) = entry {
+        let id: usize = migration_file.id.into();
```

Comment:
With the type system, we know that `migration_file.id` won’t be 0. So we could convert this ID as we do elsewhere (and have a marginally smaller vector), like this:
```diff
diff --git a/rusqlite_migration/src/loader.rs b/rusqlite_migration/src/loader.rs
index 254e71d..3b3a8b7 100644
--- a/rusqlite_migration/src/loader.rs
+++ b/rusqlite_migration/src/loader.rs
@@ -91,14 +91,14 @@ fn from(value: &MigrationFile) -> Self {
 }
 
 pub(crate) fn from_directory(dir: &'static Dir<'static>) -> Result<Vec<M<'static>>> {
-    let mut migrations: Vec<Option<M>> = vec![None; dir.dirs().count() + 1];
+    let mut migrations: Vec<Option<M>> = vec![None; dir.dirs().count()];
 
     // We cannot use FromIterator<(K, V)> for BTreeMap<K, V, Global> because that currently
     // allocates a Vec and sorts it in the background, which we want to avoid
     for dir in dir.dirs() {
         let migration_file = MigrationFile::try_from(dir)?;
 
-        let id: usize = migration_file.id.into();
+        let id: usize = migration_file.id.get() - 1;
 
         if migrations.len() <= id {
             return Err(Error::FileLoad(
@@ -122,7 +122,7 @@ pub(crate) fn from_directory(dir: &'static Dir<'static>) -> Result<Vec<M<'static
         ));
     }
 
-    if migrations.iter().skip(1).any(|m| m.is_none()) {
+    if migrations.iter().any(|m| m.is_none()) {
         return Err(Error::FileLoad(
             "Migration ids must be consecutive numbers".to_string(),
         ));
@@ -131,7 +131,6 @@ pub(crate) fn from_directory(dir: &'static Dir<'static>) -> Result<Vec<M<'static
     // The values are returned in the order of the keys, i.e. of IDs
     migrations
         .into_iter()
-        .skip(1)
         .collect::<Option<Vec<_>>>()
         .ok_or(Error::FileLoad("Could not load migrations".to_string()))
 }
```